concordium-std/src/bump_alloc.rs

```diff
@@ -0,0 +1,184 @@
+
+/// This is an invalid implementation of Sync.
+/// The [BumpAllocator] cannot be safely used from multiple threads, but that is
+/// OK since it won't be in Wasm. This [Sync] implementation is required for
+/// defining it as the global allocator.
+unsafe impl Sync for BumpAllocator {}
+
+pub struct BumpAllocator {
+    /// The next pointer to be handed out.
+    next:        UnsafeCell<usize>,
+    /// The end of the heap. Cannot be known at compile time and thus the
+    /// initial value is 0. It is updated on the first allocation.
+    heap_end:    UnsafeCell<usize>,
+    /// The number of active allocations. Used for resetting `next` to 1 if
+    /// there are no more active allocations.
+    allocations: UnsafeCell<usize>,
+    /// Stores the last address given out, and allows for a small improvement
+    /// over the always-increasing bump allocator. Namely that if an item
+    /// `a` is allocated and deallocated without any allocations in between,
+    /// then the memory of `a` is reused.
+    last_alloc:  UnsafeCell<usize>,
+}
+
+impl BumpAllocator {
+    /// Create a new [BumpAllocator].
+    ///
+    /// # Safety
+    /// - Can only be used in single-threaded environments.
+    pub const unsafe fn new() -> Self {
+        Self {
+            next:        UnsafeCell::new(MIN_PTR_ADDR),
+            heap_end:    UnsafeCell::new(0),
+            allocations: UnsafeCell::new(0),
+            // Must be set to the same initial address as `next`.
+            last_alloc:  UnsafeCell::new(MIN_PTR_ADDR),
+        }
+    }
+
+    /// Grow the memory by `delta` pages, each of size [PAGE_SIZE].
+    fn grow_memory(&self, delta: PageCount) -> PageCount {
+        // The first argument refers to the index of memory to return the size of.
+        // Currently, Wasm only supports a single slot of memory, so `0` must always be
+        // used. Source: https://doc.rust-lang.org/beta/core/arch/wasm32/fn.memory_grow.html
+        PageCount(core::arch::wasm32::memory_grow(0, delta.0))
+    }
+
+    /// Get the size of memory in terms of pages.
+    fn size(&self) -> PageCount {
+        // The argument refers to the index of memory to return the size of.
+        // Currently, Wasm only supports a single slot of memory, so `0` must always be
+        // used. Source: https://doc.rust-lang.org/beta/core/arch/wasm32/fn.memory_size.html
+        PageCount(core::arch::wasm32::memory_size(0))
+    }
+}
+
+unsafe impl GlobalAlloc for BumpAllocator {
+    unsafe fn alloc(&self, layout: Layout) -> *mut u8 {
+        let heap_end = &mut *self.heap_end.get();
+        let next = &mut *self.next.get();
+
+        // On the first allocation, the heap end is 0, as we cannot know how
+        // many pages the contract is configured to start with. So we get the
+        // actual size and update the value.
+        if *heap_end == 0 {
+            let actual_size = self.size().size_in_bytes();
+            *heap_end = actual_size;
+        }
```

Comment:
**Question:** Why is this done as part of `alloc` and not in `BumbAllocator::new`?

Reply:
I'll document that, thanks for pointing it out.
The reason is that the `new` must be a `const` method so we can declare a static with it.